### PR TITLE
BUG, TST: Several issues with scipy.signal.residue

### PR DESCRIPTION
**Pull request by nmayorov**
Hi! 

This is the partial fix of #4562, #4464, #4561. So far I worked only with `residue`. But if you generally approve I'll also take care of `residuez`.

Some notes:
1. If `k` (direct polynomial term) is zero, an empty array is returned.
2. All poles and residuals are returned, even if residuals are zero.
3. I changed `cmplx_sort` such that it always sorts by the magnitude, i.e. absolute value. This is consistent with the docstring and makes more sense from mathematical point of view. Previously it sorted complex arrays by absolute value, and real arrays by the values itself.

### CODE REVIEW

**Comment by rgommers** [comment]
Hi @nmayorov, thanks! From a quick check this looks good. The change to ``cmplx_sort`` I'll have to think about a bit - it's a backwards compatibility break so at least it needs documenting in the release notes (we can change it straight away if we deem the current behavior to be a bug though).

**Comment by pv** [comment]
Browsing through old issues...:

I think the `cmplx_sort` change breaks `unique_roots` for real inputs. 

But now that I look at `unique_roots`, the algorithm looks like nonsense for complex roots, and it's used as a subroutine in the other routines :(
```
>>> from scipy.signal.signaltools import unique_roots
>>> unique_roots([1, 1j, 1], tol=0.1)
(array([ 1.+0.j,  0.+1.j,  1.+0.j]), array([1, 1, 1]))
>>> unique_roots([1, 1, 1j], tol=0.1)
(array([ 1.+0.j,  0.+1.j]), array([2, 1]))
>>> unique_roots([1, 1, 1e-9 + 1j], tol=0.1)
(array([  1.00000000e+00+0.j,   1.00000000e-09+1.j]), array([2, 1]))
>>> unique_roots([1, 1 + 2e-9, 1e-9 + 1j], tol=0.1)
(array([  1.00000000e+00+0.j,   1.00000000e-09+1.j,   1.00000000e+00+0.j]), array([1, 1, 1]))
```

**Comment by rgommers** [comment]
Since there's a second PR for this, any update? Can we merge this somehow while avoiding breaking `unique_roots`, perhaps by having a duplicate version of `cmplx_sort`?

**Comment by nmayorov** [comment]
@rgommers 

I'll try to finish this PR without breaking other functions.

**Comment by nmayorov** [comment]
> But now that I look at unique_roots, the algorithm looks like nonsense for complex roots, and it's used as a subroutine in the other routines :(

I don't see that this algorithm was changed and yes it looks completely incorrect for complex roots. 

I will work on it then first in a separate PR.

**Comment by nmayorov** [comment]
Commit the code on to of the master.

I did change `cmplx_sort` as explained above. It does "break" `unique_roots` which is already broken, so it doesn't matter. 

Can merge this after #10961 where `unique_roots` is fixed.

**Comment by nmayorov** [comment]
@larsoner 

If the tests pass, I think it's ready to be merged. Please review the changes.

File: scipy/signal/tests/test_signaltools.py [hunk 1]
```diff
@@ -2507,6 +2507,108 @@ def test_bad_args(self):
...
+        assert_almost_equal(k, [2])
+
+        r, p, k = residue([7, 2, 3, -1], [1, -3, 2])
+        assert_almost_equal(r, [-11, 69])
+        assert_almost_equal(p, [1, 2])
+        assert_almost_equal(k, [7, 23])
+
+        r, p, k = residue([2, 3, -1], [1, -3, 4, -2])
+        order = np.argsort(p)
+        assert_almost_equal(r[order], [4, -1 + 3.5j, -1 - 3.5j])
```
Comment (larsoner) [inline scipy/signal/tests/test_signaltools.py]: 32-bit failure is real:
```
        r, p, k = residue([2, 3, -1], [1, -3, 4, -2])
        order = np.argsort(p)
>       assert_almost_equal(r[order], [4, -1 + 3.5j, -1 - 3.5j])
E       AssertionError: 
E       Arrays are not almost equal to 7 decimals
E        ACTUAL: array([-1.+3.5j, -1.-3.5j,  4.+0.j ])
E        DESIRED: [4, (-1+3.5j), (-1-3.5j)]
```

Comment (larsoner) [reply scipy/signal/tests/test_signaltools.py]: https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=3600&view=logs&jobId=dca3794b-c3dc-52e6-709a-3545ea017448

Comment (nmayorov) [reply scipy/signal/tests/test_signaltools.py]: It happens when sorting close by real part roots. Now I've made the checking algorithm very robust (with `linear_sum_assignment`.).

Now the tests should pass reliably. Merging?

**Comment by larsoner** [comment]
@nmayorov LGTM I'll go ahead and merge. Then feel free to also tackle `residuez`.

Also if this actually closes #4562, #4464, or #4561 feel free to close them

**Comment by rgommers** [comment]
Great to get this in, thanks @nmayorov

**Pull request merged**